### ISSUE TITLE
Open Project Page for http://192.168.168.25:8080/ goes to http://localhost:8080/...

### ISSUE THREAD

**Issue by Dominar** [zionyx/jenkins-tray#32]
Version Affected: 1.0.1

Issue:
- Right click on a server project "Project1" in the main Jenkins Tray window, and click on "Open Project Page" for a server who's URL is "http://192.168.168.25:8080/", and the web browser is sent to "http://localhost:8080/job/Project1/"

Impacts:
- Everyone that wants to use Jenkins Tray against a server that is not "localhost".

**Comment by zionyx** [comment]
Can you also include the application configuration file, **%AppData%\Jenkins Tray\jenkins.configuration** Thanks.

**Comment by Dominar** [comment]
Contents of jenkins.configuration:
```
{
  "servers": [
    {
      "url": "http://192.168.168.25:8080",
      "displayName": "Server1",
      "credentials": null,
      "ignoreUntrustedCertificate": false,
      "projects": [
        {
          "name": "MooMobile",
          "url": "http://localhost:8080/job/MooMobile/",
          "acknowledged": false,
          "token": null,
          "causetext": null
        }
      ]
    }
  ],
  "notificationSettings": {
    "failedSoundPath": null,
    "succeededSoundPath": null,
    "stillFailingSoundPath": null,
    "fixedSoundPath": null,
    "treatUnstableAsFailed": true
  },
  "generalSettings": {
    "refreshIntervalInSeconds": 15,
    "updateMainWindowIcon": true,
    "integrateWithClaimPlugin": true
  }
}
```

**Comment by zionyx** [comment]
Thanks for the input. I see the problem in the config file. The value of the server url is http://192.168.168.25:8080. However the value of your project url is http://localhost:8080/job/MooMobile/.

Hmm... I have to assume editing of server url with a new url which Jenkins Tray cannot reach may lead to this problem. I assume these steps can cause the problem:
1. Connect to a valid server url, eg: http://localhost:8080, and select Project1 that shows up in the server edit dialog in Jenkins Tray.
2. Jenkins Tray should be able to get some details of the project.
3. Then edit the server to change its url to http://192.168.168.25:8080, for which Jenkins Tray cannot reach.
4. Notice that the project list won't show up at all, and the status may say "Error loading projects from http://192.168.168.25:8080".
5. But Jenkins Tray still shows the status of Project1.
6. Double clicking Project1 on the main UI will bring you to http://localhost:8080/job/Project1 instead of the new server url.

If the steps above are not exactly how you get to the problem, please provide me your version.

I will look into how the cached server / project list can be improved to avoid this problem.

**Comment by Dominar** [comment]
The scenario that I'm seeing is as such.

Jenkins Server Website is at "http://192.168168.25:8080/".

Deleting all of the servers produces the following jenkins.configuration (so you know where I'm starting):
```
{
  "servers": [],
  "notificationSettings": {
    "failedSoundPath": null,
    "succeededSoundPath": null,
    "stillFailingSoundPath": null,
    "fixedSoundPath": null,
    "treatUnstableAsFailed": true
  },
  "generalSettings": {
    "refreshIntervalInSeconds": 15,
    "updateMainWindowIcon": true,
    "integrateWithClaimPlugin": true
  }
}
```

1. In "Jenkins Tray - Settings" window, click on the "Add server" button.
1. "Server settings" window will be shown.
    1. For "URL:", enter "http://192.168.168.25:8080". 
    1. For "Display name (optional):", enter "Jenkins1".
    1. Click on the "OK" button.
1. The "Servers" pane, now shows the new server entry, and the "Projects" pane shows the same projects that you can see via the Jenkins Server Website.

The newly updated jenkins.configuration file now contains the following:
```
{
  "servers": [
    {
      "url": "http://192.168.168.25:8080",
      "displayName": "Jenkins1",
      "credentials": null,
      "ignoreUntrustedCertificate": false,
      "projects": []
    }
  ],
  "notificationSettings": {
    "failedSoundPath": null,
    "succeededSoundPath": null,
    "stillFailingSoundPath": null,
    "fixedSoundPath": null,
    "treatUnstableAsFailed": true
  },
  "generalSettings": {
    "refreshIntervalInSeconds": 15,
    "updateMainWindowIcon": true,
    "integrateWithClaimPlugin": true
  }
}
```

Now, with the "Jenkins Tray - Settings" window still open, and the "Jenkins1" server still selected in the "Servers" pane, click on the checkbox for a project "TestProj Mobile", to enable it.

The newly updated jenkins.configuration file now contains the following:
```
{
  "servers": [
    {
      "url": "http://192.168.168.25:8080",
      "displayName": "Jenkins1",
      "credentials": null,
      "ignoreUntrustedCertificate": false,
      "projects": [
        {
          "name": "TestProj Mobile",
          "url": "http://localhost:8080/job/TestProj%20Mobile/",
          "acknowledged": false,
          "token": null,
          "causetext": null
        }
      ]
    }
  ],
  "notificationSettings": {
    "failedSoundPath": null,
    "succeededSoundPath": null,
    "stillFailingSoundPath": null,
    "fixedSoundPath": null,
    "treatUnstableAsFailed": true
  },
  "generalSettings": {
    "refreshIntervalInSeconds": 15,
    "updateMainWindowIcon": true,
    "integrateWithClaimPlugin": true
  }
}
```

Note the difference in the hostname for the Server URL and the Project URL that were generated to the configuration file.

If I substitute "192.168.168.25" for "localhost" in the Project URL, I get "http://192.168.168.25:8080/job/TestProj%20Mobile/". I can then web browse to that URL to visit the Jenkins page for that specific project.

**Comment by Dominar** [comment]
Grrr. I took a look at your Logs/general.txt file output. It turns out, that the project URL plus the "/api/xml" appended onto the end of the URL, is received XML data back from Jenkins. However, the data in the XML, is where the "localhost" domain references are probably originating.

That being the case, I no longer think this is a bug, but is somehow a misconfiguration of Jenkins. I'm trying to confirm this on this end.

If that is indeed the case, then please accept my apologies. However, if at that time you were then able to implement detection of localhost references coming back in /api/xml data for a Jenkins server that is hosted on a non-localhost machine, that would be most helpful.

**Comment by Dominar** [comment]
It's official. This is the result of a misconfiguration in Jenkins. (shown below)

![2015-05-13_14-06-14](https://cloud.githubusercontent.com/assets/4212972/7617482/570b35dc-f979-11e4-8684-2da1e69022d2.gif)

I think detection of this misconfiguration, would definitely be helpful, and probably be very easy to implement. The Project URLs should only contain localhost, when the Server URLs contain localhost. Anything else, is probably an acceptable scenario.

I'm sorry for any inconvenience.

**Comment by zionyx** [comment]
That makes sense. Thanks alot for the details.

I will change the issue summary and implement this as an enhancement to detect Jenkins server misconfiguration prompt. 

I have also created a new issue for #33.

**Comment by zionyx** [comment]
Proposed notification enhancement:
* Notify user about server url and project url difference on server add / edit.
* Notify user every time when the server edit dialog tries to load projects from the server.
* Allow this notification to be silenced by having a checkbox in the notification dialog.
* Notification silence is saved to the config file as a server setting.
* Have not thought of any way to allow users to clear this flag other than manually removing it from jenkins.configuration file.
* Try to test against Jenkins Enterprise whenever possible to see if it breaks this implementation.
* Sample message: Project url does not match the server url domain name. Please check if Jenkins is misconfigured as ________.

**Comment by zionyx** [comment]
Hi @Dominar, i still cannot properly reproduce your problem despite the misconfigured Jenkins master server. My own server is also configured as localhost but server url + /api/xml returns the 192.168.0.100 ip address, the same ip as i access it via a browser, for instance http://192.168.0.100:8080/.

I will produce a trial fix and let you try it out, if that is alright with you. Thanks.

**Comment by Dominar** [comment]
Yes, I'll put it through the paces on this end, since I'm able to reproduce
to localhost Jenkins issue. By the way, we're using Jenkins ver. 1.608, on
Microsoft Windows. If that somehow matters. It shouldn't, but just in case.
=)

**Comment by zionyx** [comment]
Cool. I tested on Windows with version 1.613 without avail. 
I will let you know when the feature branch is ready for testing. Thanks.
It may take a while though. Cheers.

**Comment by zionyx** [comment]
Hey @Dominar, I can finally reproduce this problem. I'll work on it soon.
